{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "# Introduction\n",
    "\n",
    "This notebook we explore pancreatic islets cell type composition. First, we provide a guide to using Xenium data and transferring cell type annotations from a reference dataset using spatial transcriptomics data tools. We demonstrate how to load, preprocess, and analyze multimodal imaging data, enabling exploration of spatially resolved cell types and their associated features. The workflow includes downloading datasets, preparing spatial and reference data, applying dimensionality reduction techniques like PCA and UMAP, and transferring annotations. Visualization methods are showcased, including spatial plots with cell boundaries and phenotypic markers, focusing on specific regions and subsets of cell types. Ultimately we search for any spatial neighbors of endocrine cells. We process endocine cells connected to endocrine cells and determine their groups (i.e., islets). The analysis allows to see relative composition of islets by cell type, size and presence of non-endocrine cell types within and around islets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from scipy.cluster import hierarchy\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage, fcluster\n",
    "from scipy.spatial.distance import cdist, pdist, squareform\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Polygon\n",
    "from matplotlib.collections import PatchCollection\n",
    "from matplotlib.lines import Line2D\n",
    "import matplotlib.cm as cm\n",
    "import tifffile\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Transfer cell type annotation from one dataset to another\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get started, download two datasets (~3 GB and ~0.4 GB) from the provided Zenodo repository, then extract (unpack) \n",
    "the contents to a local directory on your machine. After unpacking, locate the variable named dataPath \n",
    "in the cell below and update its value to point to the folder where you extracted the data. Once you've \n",
    "set the correct path, you can run the subsequent cells to visualize \n",
    "and explore the multimodal imaging data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!curl -L \"https://zenodo.org/api/records/15777497/files-archive\" -o dataset.zip && unzip dataset.zip && rm dataset.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!curl -L \"https://zenodo.org/api/records/15777586/files-archive\" -o dataset.zip && unzip dataset.zip && rm dataset.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dataPath = './'\n",
    "dataPath = '/projects/activities/kappsen-tmc/USERS/domans/examples-sennet-portal/'\n",
    "\n",
    "# Load the annotated reference dataset\n",
    "adata = sc.read_h5ad(dataPath + 'x-subset-50k-all-umap-annotated.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Palette for pancreas cell types\n",
    "palette_pancreas_fine = {'Acinar-1': 'orchid', 'Acinar-2': 'blue',\n",
    "                        'Ductal': 'deepskyblue', 'Ductal-Other-1': 'crimson', 'Ductal-Other-2': 'teal',\n",
    "                        'Endocrine-Alpha': 'yellow', 'Endocrine-Beta': 'orange', 'Endocrine-Delta': 'limegreen', 'Endocrine-Gamma': 'orangered',\n",
    "                        'Endocrine-Other-1': 'purple', 'Endocrine-Other-2': 'darkgreen',\n",
    "                        'Endothelial': 'chocolate', 'Pericytes': 'turquoise', 'Fibroblasts': 'maroon', 'Adipocytes': 'gold',\n",
    "                        'Lymphoid': 'dimgrey', 'Mast': 'deeppink', 'Myeloid': 'lime',\n",
    "                        'Neural-1': 'cyan', 'Neural-2': 'black', 'Neural-3': 'cornflowerblue'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_xenium_dataset(id, path, N=None, suffix=None, f=1.):\n",
    "\n",
    "    \"\"\"Load a Xenium dataset from a specified Xenium bundle directory.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    id : str\n",
    "        Identifier for the sample (not used internally).\n",
    "    path : str\n",
    "        Path to the Xenium bundle directory. This directory should contain\n",
    "        'cells.parquet' and 'cell_feature_matrix.h5' files.\n",
    "    N : int, optional\n",
    "        Number of cells to randomly sample. If None, use all cells.\n",
    "    suffix : str, optional\n",
    "        Suffix to append to cell indices.\n",
    "    f : float, optional\n",
    "        Scaling factor for spatial coordinates (default is 1.0).\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    adata : AnnData\n",
    "        Data object with spatial coordinates in `adata.obsm['spatial']`.\n",
    "    \"\"\"\n",
    "\n",
    "    obs = pd.read_parquet(path + '/cells.parquet', engine='auto', columns=None,\n",
    "                          storage_options=None, use_nullable_dtypes=False).set_index('cell_id')\n",
    "    adata = sc.read_10x_h5(path + '/cell_feature_matrix.h5')\n",
    "    if not N is None:\n",
    "        adata = adata[adata.obs.sample(N).index]\n",
    "    adata.obs = obs.loc[adata.obs.index]\n",
    "    if not suffix is None:\n",
    "        adata.obs.index = adata.obs.index + suffix\n",
    "    adata.obsm['spatial'] = (adata.obs[['x_centroid', 'y_centroid']]*f).astype(int).values\n",
    "\n",
    "    return adata\n",
    "\n",
    "def load_spatial_annotated(id, path, adata_ref, lowf=5, highf=10,\n",
    "                        thumb_quantile=0.99, cell_diameter_fullres=7, mpp=0.2125):\n",
    "\n",
    "    \"\"\"Load a Xenium dataset and annotate it with spatial information and cell types.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    id : str\n",
    "        Identifier for the sample.\n",
    "\n",
    "    path : str\n",
    "        Path to the Xenium bundle directory. This directory should contain\n",
    "        'cells.parquet', 'cell_feature_matrix.h5', and other relevant files.\n",
    "\n",
    "    adata_ref : AnnData\n",
    "        Reference AnnData object containing cell type annotations.\n",
    "\n",
    "    lowf : int, optional\n",
    "        Downsampling factor for low-resolution images (default is 5).\n",
    "\n",
    "    highf : int, optional\n",
    "        Downsampling factor for high-resolution images (default is 10).\n",
    "\n",
    "    thumb_quantile : float, optional\n",
    "        Quantile for thumbnail intensity scaling (default is 0.99).\n",
    "\n",
    "    cell_diameter_fullres : int, optional\n",
    "        Diameter of cells in full resolution (default is 7).\n",
    "\n",
    "    mpp : float, optional\n",
    "        Microns per pixel (default is 0.2125).\n",
    "    \"\"\"\n",
    "\n",
    "    ad = load_xenium_dataset(id, path, suffix='.' + id)\n",
    "\n",
    "    print('Loading PCA data...')\n",
    "    with open(f'{path}/PCA-reference.pklz', 'rb') as tempfile:\n",
    "        xgenes, tr_PCA = pickle.load(tempfile)\n",
    "    ad.obsm['X_pca'] = tr_PCA.transform(ad[:, xgenes].X)\n",
    "\n",
    "    print('Loading UMAP coordinates...')\n",
    "    if True:\n",
    "        # Load precomputed UMAP coordinates\n",
    "        ad.obsm['X_umap'] = pd.read_parquet(f'{path}/UMAP.parquet', engine='auto', columns=None).loc[ad.obs.index].values\n",
    "    else:\n",
    "        # Use UMAP from reference could take under 1 hour\n",
    "        # and the UMAP-reference.pklz is 7 GB, not shared\n",
    "        with open(f'{path}/UMAP-reference.pklz', 'rb') as tempfile:\n",
    "            tr_UMAP = pickle.load(tempfile)\n",
    "        ad.obsm['X_umap'] = pd.DataFrame(index=ad.obs.index, data=tr_UMAP.transform(ad.obsm['X_pca'])).fillna(20.).values\n",
    "\n",
    "    print('Log-transforming data...')\n",
    "    sc.pp.log1p(ad)\n",
    "\n",
    "    print('Loading thumbnail image data...')\n",
    "    if True:\n",
    "        # Load subsampled image\n",
    "        thumbnail = tifffile.imread(f'{path}/thumbnail.tiff')\n",
    "    else:\n",
    "        # Load full image image\n",
    "        imagef = f'{path}/morphology_focus/morphology_focus_0000.ome.tif'\n",
    "        thumbnail = tifffile.imread(imagef)[::highf, ::highf]\n",
    "        tifffile.imwrite(f'{path}/thumbnail.tiff', thumbnail)\n",
    "\n",
    "    thumbnail = (255. * thumbnail / np.quantile(thumbnail.ravel(), thumb_quantile))\n",
    "    thumbnail[thumbnail>255] = 255\n",
    "    thumbnail = 255 - thumbnail\n",
    "    thumbnail = (np.dstack([thumbnail]*3)).astype(np.uint8)\n",
    "\n",
    "    print(f'Setting up spatial data for thunbnail {thumbnail.shape}...')\n",
    "    ad.uns['spatial'] = {'library_id': {'images': {'hires': thumbnail,\n",
    "                                        'lowres': thumbnail[::lowf, ::lowf]},\n",
    "                        'scalefactors': {'tissue_hires_scalef': 1. / (mpp * highf),\n",
    "                                        'tissue_lowres_scalef': 1. / (mpp * lowf * highf),\n",
    "                                        'spot_diameter_fullres': cell_diameter_fullres}}}\n",
    "\n",
    "    print('Loading mIF data...')\n",
    "    codex = pd.read_parquet(f'{path}/mIF-cells.parquet', engine='auto', columns=None)\n",
    "    codex = codex.drop(['centroid-0', 'centroid-1'], axis=1)\n",
    "    codex.index = codex.index + '.' + id\n",
    "    codex = codex.reindex(ad.obs.index).fillna(0.)\n",
    "    ad.obs[codex.columns] = codex.values\n",
    "\n",
    "    def annotate_sh(ind, df, df_ref):\n",
    "        return pd.Series(index=ind, data=df_ref.index[cdist(df, df_ref, metric='correlation').argmin(axis=1)])\n",
    "\n",
    "    print('Transferring cell type annotations...')\n",
    "    identity = 'Celltype fine'\n",
    "    df_sh_ref_fine = adata_ref.to_df()\n",
    "    df_sh_ref_fine.index = adata.obs[identity].reset_index(drop=True).values\n",
    "    df_sh_ref_fine = df_sh_ref_fine.groupby(level=0).mean()\n",
    "\n",
    "    ad.obs[identity] = annotate_sh(ad.obs.index, ad.to_df(), df_sh_ref_fine)\n",
    "\n",
    "    ad.uns['df_cell_boundaries'] = pd.read_parquet(f'{path}/cell_boundaries.parquet')\n",
    "\n",
    "    return ad\n",
    "\n",
    "def plotWithBoundaries(ad, x1, x2, y1, y2, var=None, dfb=None, palette='tab20',\n",
    "                    figsize=(7, 7), colormap={}, colormap2='viridis', s='', exclusion='',\n",
    "                    vmin=None, vmax=None, qmin=None, qmax=None, invert=False, title=''):\n",
    "\n",
    "    \"\"\"Plot cell boundaries with color coding based on a specified variable.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    ad : AnnData\n",
    "        Annotated data object containing spatial coordinates and cell type annotations.\n",
    "\n",
    "    x1, x2, y1, y2 : float\n",
    "        Coordinates defining the region of interest to plot.\n",
    "\n",
    "    var : str, optional\n",
    "        Variable to use for color coding the cell boundaries.\n",
    "\n",
    "    dfb : pd.DataFrame, optional\n",
    "        DataFrame containing cell boundary information. If None, it will be loaded from `ad.uns['df_cell_boundaries']`.\n",
    "    \n",
    "    palette : str, optional\n",
    "        Name of the color palette to use for cell types (default is 'tab20').\n",
    "\n",
    "    figsize : tuple, optional\n",
    "        Size of the figure to create (default is (7, 7)).\n",
    "\n",
    "    colormap : dict, optional\n",
    "        Custom colormap for cell types. If provided, it will override the default palette.\n",
    "\n",
    "    colormap2 : str, optional\n",
    "        Name of the colormap to use for continuous variables (default is 'viridis').\n",
    "\n",
    "    s : str, optional\n",
    "        Suffix to append to cell indices in the DataFrame (default is '').\n",
    "\n",
    "    exclusion : str, optional\n",
    "        String to exclude from the labels (default is '').\n",
    "\n",
    "    vmin, vmax : float, optional\n",
    "        Minimum and maximum values for color scaling. If None, they will be determined from the data.\n",
    "\n",
    "    qmin, qmax : float, optional\n",
    "        Quantiles to use for determining vmin and vmax. If both are None, the full range of the variable will be used.\n",
    "\n",
    "    invert : bool, optional\n",
    "        If True, invert the y-axis (default is False).\n",
    "\n",
    "    title : str, optional\n",
    "        Title of the plot (default is an empty string).\n",
    "    \"\"\"\n",
    "    colormap_ = colormap.copy()\n",
    "\n",
    "    if var in ad.obs:\n",
    "        se = ad.obs[var]\n",
    "    elif var in ad.var.index:\n",
    "        se = ad[:, var].to_df()[var]\n",
    "    else:\n",
    "        raise ValueError(f'Variable {var} not found in adata.obs or adata.var.index')\n",
    "\n",
    "    if (se.dtype == 'category') or (se.dtype == bool) or (se.dtype == str) or (se.dtype == 'O'):\n",
    "        se_max = None\n",
    "    else:\n",
    "        se_max = se.max()\n",
    "\n",
    "    if not qmin is None and not qmax is None:\n",
    "        vmin = se.quantile(qmin)\n",
    "        vmax = se.quantile(qmax)\n",
    "\n",
    "    def calculate_legend_columns(num_items, max_height, item_height):\n",
    "        max_items_per_column = max_height // item_height\n",
    "        num_columns = (num_items + max_items_per_column - 1) // max_items_per_column\n",
    "        return num_columns\n",
    "\n",
    "    dfb_sub = dfb[(dfb['vertex_x']>=x1) & (dfb['vertex_x']<=x2) & (dfb['vertex_y']>=y1) & (dfb['vertex_y']<=y2)]\n",
    "\n",
    "    gb = dfb_sub[['cell_id', 'vertex_x', 'vertex_y']].groupby('cell_id').agg(list)\n",
    "    dgb = {}\n",
    "    for c in gb.index:\n",
    "        v = gb.loc[c].values\n",
    "        v = np.vstack([np.array(v[0]), np.array(v[1])]).T\n",
    "        dgb[c] = v\n",
    "    print(f'Found {len(dgb)} cell boundaries in the specified region.')\n",
    "\n",
    "    backup_palette = plt.get_cmap(palette)\n",
    "    N_colors = backup_palette.N\n",
    "    used_color_index = 0\n",
    "\n",
    "    patches = []\n",
    "    keys = list(dgb.keys())\n",
    "    colors = {}\n",
    "    labels = {}\n",
    "    ucolors = {}\n",
    "    for c in keys:\n",
    "        if c + '.' + s in se.index:\n",
    "            if se_max is None:\n",
    "                celltype = se.loc[c + '.' + s]\n",
    "                if not celltype in colormap_:\n",
    "                    colormap_[celltype] = backup_palette(used_color_index % N_colors)\n",
    "                    used_color_index += 1\n",
    "                colors[c] = colormap_[celltype]\n",
    "                labels[c] = celltype\n",
    "                if celltype not in ucolors:\n",
    "                    ucolors[celltype] = colors[c]\n",
    "\n",
    "    for c in keys:\n",
    "        if c + '.' + s in se.index:\n",
    "            v = dgb[c]\n",
    "            if se_max is None:\n",
    "                patches.append(Polygon(v, closed=True, fill=True,\n",
    "                                    edgecolor='k' if not exclusion in labels[c] else colors[c],\n",
    "                                    linewidth=0.5, facecolor=colors[c], label=labels[c]))\n",
    "            else:\n",
    "                cmap = plt.get_cmap(colormap2, 256)\n",
    "                vc = se.loc[c + '.' + s]\n",
    "                if vmin is not None and vmax is not None:\n",
    "                    vc = (vc - vmin) / (vmax - vmin)\n",
    "                    vc = np.clip(vc, 0, 1)\n",
    "                patchcolor = cmap(vc)\n",
    "                patches.append(Polygon(v, closed=True, fill=True,\n",
    "                                    edgecolor='k' if not exclusion=='' else patchcolor,\n",
    "                                    linewidth=0.5, facecolor=patchcolor, label=None))\n",
    "\n",
    "    unique_patches = {patch.get_label(): patch for patch in patches}.values()\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "    pc = PatchCollection(patches, match_original=True)\n",
    "    ax.add_collection(pc)\n",
    "    ax.set_xlim(x1, x2)\n",
    "    ax.set_ylim(y2, y1)\n",
    "    ax.set_xlabel('X Coordinate, µm', fontsize=14)\n",
    "    ax.set_ylabel('Y Coordinate, µm', fontsize=14)\n",
    "    ax.set_aspect('equal')\n",
    "    if invert:\n",
    "        ax.invert_yaxis()\n",
    "    plt.title(title, fontsize=16)\n",
    "\n",
    "    if se_max is None:\n",
    "        max_height = fig.get_figheight() * fig.dpi\n",
    "        item_height = 25\n",
    "        num_columns = calculate_legend_columns(len(unique_patches), max_height, item_height)\n",
    "\n",
    "        cic_handles = []\n",
    "        mkeys = sorted(list(ucolors.keys()), key=str.lower)\n",
    "        for celltype in mkeys:\n",
    "            color = ucolors[celltype]\n",
    "            cic_handles.append(Line2D([0], [0], marker='o', color='w', label=celltype, markerfacecolor=color, markersize=12))\n",
    "        ax.legend(handles=cic_handles, bbox_to_anchor=(1, 0.5), loc='center left', ncol=num_columns, fancybox=False,\n",
    "                frameon=False, fontsize=12, title='', title_fontsize=16)\n",
    "    else:\n",
    "        pc.set_cmap(colormap2)\n",
    "        pc.set_clim(vmin=vmin, vmax=vmax)\n",
    "        cbar = plt.colorbar(pc, ax=ax, label=var, orientation='vertical', fraction=0.03, pad=0.04)\n",
    "        cbar.ax.set_ylabel(var, fontsize=14)\n",
    "        cbar.ax.tick_params(labelsize=12)\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "id = 'JDC-WP-012-w'\n",
    "# dataPath = './'\n",
    "dataPath = '/projects/activities/kappsen-tmc/USERS/domans/examples-sennet-portal/JDC-WP-012-w-xenium-slim/'\n",
    "\n",
    "# Load the spatial annotated dataset example, containing all cells for the specified ID\n",
    "ad = load_spatial_annotated(id, dataPath, adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variables to explore\n",
    "\n",
    "**Genes**:\n",
    "ABCC8, ACKR1, ACKR3, ACSL1, ACTA2, ADCYAP1, ADIPOQ, ALDH1A1, AMY2A, ANKRD33B, AQP1, AQP8, ARX, ASCL1, AURKA, BARX2, BAX, BCL2, BCL2L1, BCL2L2, BUB1, C11orf96, C1QA, C3, CALCA, CALCB, CALD1, CARTPT, CASR, CAVIN1, CCL2, CCL5, CCNB2, CCND1, CCNE2, CCNL2, CCR2, CD2, CD274, CD3D, CD3E, CD3G, CD52, CD55, CD63, CD68, CD7, CD74, CD84, CD8A, CD9, CD96, CDC20, CDH19, CDK1, CDK5RAP3, CDKN1A, CDKN1C, CDKN2A, CDX2, CEACAM6, CFTR, CHGA, CHGB, CHRM3, CLDN18, CLU, CMTM8, COL1A1, COL1A2, COL6A2, COMP, CORT, CPA3, CREB5, CRHR2, CSF1R, CTNNB1, CTRB1, CTSB, CTSK, CXCL1, CXCL12, CXCL2, CXCL6, CXCL8, CXCR4, DDX60L, DLK1, DPT, E2F1, ENO1, ERO1B, ERP27, ESAM, ETV1, FABP4, FAP, FDXR, FFAR3, FFAR4, FOLR1, FOSB, FOXB1, FOXB2, FOXO1, FOXP3, FRZB, G6PC2, GABRA6, GAL, GATM, GC, GCG, GCGR, GCNT3, GDF15, GEM, GHRL, GHSR, GLIS3, GLP1R, GLS, GPM6B, GPNMB, GPX3, GSDMB, GSDMD, GSDME, GSN, GZMK, HADH, HES1, HES4, HHEX, HLA-E, HNF1B, HPGDS, HSF4, HSPB1, IAPP, IFNG, IGFBP7, IGHD, IGHG1, IGHM, IL1RL1, IL2RB, IL32, INS, IRX2, ISL1, JUN, JUND, KIT, KLRC1, KLRG1, KRAS, KRT19, KRT7, KRT8, L1CAM, LCN2, LENG8, LEPR, LGALS1, LGI4, LGMN, LOXL4, LPL, LRFN5, LYZ, MAFA, MALL, MAP1B, MCM6, MDM2, MEIS2, MLXIPL, MMP7, MRC1, MS4A8, MT1A, MUC13, MUC5AC, MUC5B, MYC, NAP1L4, NEUROG1, NEUROG3, NKX2-2, NKX6-1, NPTX2, NPY, NPY1R, NR5A2, NUSAP1, ONECUT1, OPN4, PCNA, PDGFRA, PDGFRB, PDPN, PDX1, PECAM1, PGC, PGF, PHGR1, PHIP, PHLDA3, PLIN2, PLK3, PLXNB2, POLD3, PPP1R1B, PPY, PRDX1, PRG4, PROX1, PRSS1, PTF1A, PYY, RB1, RBP4, RBPJL, RGS5, S100A11, S100A6, SAMSN1, SCNN1G, SCTR, SDC1, SDS, SERPINA1, SERPINE1, SERPINH1, SERPINI2, SIGLEC6, SIX2, SIX3, SLC22A6, SLC30A8, SLC6A4, SLC6A6, SLC7A2, SLC7A8, SNRNP70, SORL1, SOX10, SOX4, SOX9, SPARC, SPATA2L, SPHK1, SPIC, SPINK1, SPINT2, SQSTM1, SSR1, SST, SSTR1, SSTR2, STMN2, STX1A, SULT1C2, SYCN, SYT13, TAC1, TAGLN, TEN1, TFF1, TFF2, TGFB1I1, TGFBR2, TIMP1, TM4SF1, TM4SF4, TMSB10, TNFRSF10B, TNFRSF12A, TNFRSF1A, TP53, TRIAP1, TSPAN1, TSPAN8, TTR, TUBGCP2, TXNIP, UBALD2, UCHL1, UCP1, VEGFA, VGF, VIM, VIP, VWA5A, XCL1, YBX3, ZBTB7A, ZMAT3\n",
    "\n",
    "In addition to the 300 genes the data contains cell type annotation (**Celltype fine**), number of transcripts per cell (**total_counts**), cell size (**cell_area**), <br>\n",
    "and 39 channels of **PhenoCycler** data (a.k.a., CODEX): <br>\n",
    "DAPI, CD45RA, S100B, <br>\n",
    "Insulin, Carbonic Anhydrase 9, Vimentin, <br>\n",
    "Glucagon, CD45RO, Collagen I, p21, <br>\n",
    "MPO, MECOM, CFTR, PGP9.5, <br>\n",
    "Keratin 19, 53BP1, p16, Cadherin 11, <br>\n",
    "Somatostatin, CD8, Lamin B1, CD3e, <br>\n",
    "C-Peptide, HMGB1, Perilipin, Podoplanin, <br>\n",
    "Pancreatic Polypeptide, CD4, E-cadherin, <br>\n",
    "CD31, CD68, Ki67, EpCAM, CD45, <br>\n",
    "CD56, CD20, FOXP3, gH2AX, SMA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Spatial plot without boundaries, cells are shown as dots colored by cell type\n",
    "sc.pl.spatial(ad, color=['Celltype fine'], frameon=False, size=1, palette=palette_pancreas_fine, title='')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "# Date: 2025-06-06\n",
    "# Organization: The Jackson Laboratory\n",
    "# Author: Sergii Domanskyi\n",
    "\n",
    "Description: This module provides functions to identify cells of interest based on their neighboring cell types\n",
    "using a KDTree for efficient spatial queries. It also includes a function to visualize the results on a spatial layout.\n",
    "\n",
    "Module spatialsearch is a part of spatial-omics-tools:\n",
    "https://github.com/TheJacksonLaboratory/spatial-omics-tools\n",
    "\"\"\"\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "from scipy.spatial import KDTree\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import defaultdict, deque\n",
    "\n",
    "def find(df: pd.DataFrame, radius: int, threshold: float,\n",
    "                celltype_column: str, x_column: str, y_column: str,\n",
    "                interest_celltypes: list, neighbor_celltypes: list,\n",
    "                verbose: bool=False) -> pd.Index:\n",
    "\n",
    "    \"\"\"Function to identify cells of interest based on their neighbors.\n",
    "    Utilizes a KDTree for efficient spatial queries.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        DataFrame containing cell data with columns for cell types and coordinates.\n",
    "\n",
    "    radius : int\n",
    "        Radius in micrometers to search for neighboring cells.\n",
    "\n",
    "    threshold : float\n",
    "        Proportion threshold of neighboring cells that must be of the specified types.\n",
    "\n",
    "    celltype_column : str\n",
    "        Name of the column in `df` that contains cell type information.\n",
    "\n",
    "    x_column : str\n",
    "        Name of the column in `df` that contains the x-coordinates of cells.\n",
    "\n",
    "    y_column : str\n",
    "        Name of the column in `df` that contains the y-coordinates of cells.\n",
    "\n",
    "    interest_celltypes : list\n",
    "        List of cell types to consider as 'interest' cells, e.g., Endocrine.\n",
    "\n",
    "    neighbor_celltypes : list\n",
    "        List of cell types that are considered as neighbors.\n",
    "\n",
    "    verbose : bool, optional\n",
    "        Whether to print progress and statistics, default is True.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.Index\n",
    "        Index of cells that meet the criteria of having a sufficient \n",
    "        proportion of specified neighbor cell types.\n",
    "    \"\"\"\n",
    "\n",
    "    interest_cells = df[df[celltype_column].isin(interest_celltypes)]\n",
    "\n",
    "    tree = KDTree(df[[x_column, y_column]].values)\n",
    "\n",
    "    a = df[celltype_column].isin(neighbor_celltypes).values\n",
    "    idx = df.columns.get_loc(x_column)\n",
    "    idy = df.columns.get_loc(y_column)\n",
    "\n",
    "    found = []\n",
    "    sizes = []\n",
    "    with tqdm(total=len(interest_cells), desc='Searching', disable=not verbose) as pbar:\n",
    "        for i, (index, row) in enumerate(interest_cells.iterrows()):\n",
    "            if i % 10**3 == 0:\n",
    "                pbar.update(10**3)\n",
    "\n",
    "            neighbors_idx = tree.query_ball_point([row.values[idx], row.values[idy]], radius)\n",
    "            sizes.append(len(neighbors_idx))\n",
    "            proportion = a[neighbors_idx].mean()\n",
    "            if proportion > threshold:\n",
    "                found.append(index)\n",
    "\n",
    "    if verbose:\n",
    "        print('Average number of neighbors:', int(np.mean(sizes)))\n",
    "        print('Number of cells found:', len(found))\n",
    "\n",
    "    return pd.Index(found)\n",
    "\n",
    "def find_neighbors(df: pd.DataFrame, radius: int,\n",
    "                x_column: str, y_column: str,\n",
    "                cells_of_interest: list,\n",
    "                verbose: bool=False) -> pd.Index:\n",
    "\n",
    "    \"\"\"Function to find lists of neighbor cells for each cell of interest.\n",
    "    Utilizes a KDTree for efficient spatial queries.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        DataFrame containing cell data with columns for cell types and coordinates.\n",
    "\n",
    "    radius : int\n",
    "        Radius in micrometers to search for neighboring cells.\n",
    "\n",
    "    x_column : str\n",
    "        Name of the column in `df` that contains the x-coordinates of cells.\n",
    "\n",
    "    y_column : str\n",
    "        Name of the column in `df` that contains the y-coordinates of cells.\n",
    "\n",
    "    cells_of_interest : list\n",
    "        List of cell indices for which to find neighbors.\n",
    "\n",
    "    verbose : bool, optional\n",
    "        Whether to print progress and statistics, default is True.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    dict\n",
    "        Dictionary where keys are indices of interest cells and\n",
    "        values are lists of indices of neighboring cells.\n",
    "    \"\"\"\n",
    "\n",
    "    interest_cells = df.loc[cells_of_interest]\n",
    "\n",
    "    tree = KDTree(df[[x_column, y_column]].values)\n",
    "\n",
    "    idx = df.columns.get_loc(x_column)\n",
    "    idy = df.columns.get_loc(y_column)\n",
    "\n",
    "    found = {}\n",
    "    with tqdm(total=len(interest_cells), desc='Searching', disable=not verbose) as pbar:\n",
    "        for i, (index, row) in enumerate(interest_cells.iterrows()):\n",
    "            if i % 10**3 == 0:\n",
    "                pbar.update(10**3)\n",
    "\n",
    "            neighbors_idx = tree.query_ball_point([row.values[idx], row.values[idy]], radius)\n",
    "            found.update({index: df.index[neighbors_idx].values.tolist()})\n",
    "\n",
    "    return found\n",
    "\n",
    "def show(df: pd.DataFrame, cells: pd.Index,\n",
    "        title: str, figsize=(8, 4), s: float=0.1,\n",
    "        x_column: str='x_centroid', y_column: str='y_centroid',\n",
    "        colors: list=['lightgrey', 'crimson'], labels: list=['Other', 'Found'],\n",
    "        bounds=None, axes=False, legend=True, returnFig=False):\n",
    "\n",
    "    \"\"\"Function to visualize the identified cells on a sptial layout plot.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        DataFrame containing cell data with columns for coordinates.\n",
    "\n",
    "    cells : pd.Index\n",
    "        Index of cells to highlight in the plot.\n",
    "\n",
    "    title : str\n",
    "        Title of the plot.\n",
    "\n",
    "    figsize : tuple, optional\n",
    "        Size of the figure, default is (8, 4).\n",
    "\n",
    "    s : float, optional\n",
    "        Size of the scatter points, default is 0.1.\n",
    "\n",
    "    x_column : str\n",
    "        Name of the column in `df` that contains the x-coordinates of cells.\n",
    "\n",
    "    y_column : str\n",
    "        Name of the column in `df` that contains the y-coordinates of cells.\n",
    "\n",
    "    colors : list, optional\n",
    "        List of colors for plotting, default is ['lightgrey', 'crimson'].\n",
    "\n",
    "    labels : list, optional\n",
    "        List of labels for the legend, default is ['Other', 'Found'].\n",
    "\n",
    "    bounds : tuple, optional\n",
    "        Bounds for the x and y axes in the format (xmin, xmax, ymin, ymax), default is None.\n",
    "\n",
    "    axes : bool, optional\n",
    "        Whether to show axes in the plot, default is False.\n",
    "\n",
    "    legend : bool, optional\n",
    "        Whether to display the legend, default is True.\n",
    "    \"\"\"\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "    plt.title(title, fontsize=20)\n",
    "\n",
    "    if bounds is not None:\n",
    "        ax.set_xlim(bounds[0], bounds[1])\n",
    "        ax.set_ylim(bounds[2], bounds[3])\n",
    "    \n",
    "    ax.scatter(df[x_column], df[y_column], s=s, c=colors[0], alpha=1., label=labels[0])\n",
    "    ax.scatter(df.loc[cells, x_column], df.loc[cells, y_column], s=s, c=colors[1], label=labels[1])\n",
    "\n",
    "    ax.set_xlabel('X Centroid (µm)', fontsize=16)\n",
    "    ax.set_ylabel('Y Centroid (µm)', fontsize=16)\n",
    "    plt.xticks(fontsize=14, rotation=90)\n",
    "    plt.yticks(fontsize=14)\n",
    "    if legend:\n",
    "        plt.legend(frameon=False, fontsize=14, markerscale=15,\n",
    "                loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "    ax.set_aspect('equal')\n",
    "    ax.invert_yaxis()\n",
    "    plt.tight_layout()\n",
    "    if not axes:\n",
    "        plt.axis('off')\n",
    "\n",
    "    if returnFig:\n",
    "        return fig\n",
    "    else:\n",
    "        plt.show()\n",
    "\n",
    "    return\n",
    "\n",
    "def find_connected_components_with_threshold(graph_dict, min_connections=2):\n",
    "\n",
    "    \"\"\"Find connected components in a graph represented as an adjacency list.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    graph_dict : dict\n",
    "        Dictionary representing the graph where keys are nodes and values are lists of neighboring nodes.\n",
    "\n",
    "    min_connections : int, optional\n",
    "        Minimum number of connections required to keep a component together (default is 2).\n",
    "    \"\"\"\n",
    "\n",
    "    # Build bidirectional adjacency list and collect all nodes\n",
    "    adj, nodes = defaultdict(set), set()\n",
    "    for n, nbrs in graph_dict.items():\n",
    "        nodes.update([n] + nbrs)\n",
    "        for nbr in nbrs:\n",
    "            adj[n].add(nbr), adj[nbr].add(n)\n",
    "    \n",
    "    # Find connected components using BFS\n",
    "    visited, components = set(), []\n",
    "    for node in nodes:\n",
    "        if node in visited: continue\n",
    "        comp, q = [], deque([node])\n",
    "        visited.add(node)\n",
    "        while q:\n",
    "            curr = q.popleft()\n",
    "            comp.append(curr)\n",
    "            for nbr in adj[curr]:\n",
    "                if nbr not in visited:\n",
    "                    visited.add(nbr), q.append(nbr)\n",
    "        components.append(comp)\n",
    "    \n",
    "    # Split components with insufficient connections\n",
    "    final = []\n",
    "    for comp in components:\n",
    "        if len(comp) <= 1 or sum(1 for n in comp for nbr in adj[n] if nbr in comp and n < nbr) >= min_connections:\n",
    "            final.append(comp)\n",
    "        else:\n",
    "            final.extend([[n] for n in comp])\n",
    "    \n",
    "    return {f'C{i}': sorted(comp) for i, comp in enumerate(final)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Search for Endocrine cells connected to endocrine cells and determine their groups (i.e., islets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "endocrine_subtypes = ['Endocrine-Alpha', 'Endocrine-Beta', 'Endocrine-Delta', 'Endocrine-Gamma', 'Endocrine-Other-1', 'Endocrine-Other-2']\n",
    "endocrine_cells = ad.obs.index[ad.obs['Celltype fine'].isin(endocrine_subtypes)]\n",
    "found = find_neighbors(ad.obs, radius=16, x_column='x_centroid', y_column='y_centroid', cells_of_interest=endocrine_cells, verbose=True)\n",
    "\n",
    "# Filter the found neighbors to only include those that are also endocrine cells\n",
    "for cell in found.keys():\n",
    "    found[cell] = [c for c in found[cell] if c in endocrine_cells]\n",
    "\n",
    "# Find islets as connected components with at least 5 connections\n",
    "result = find_connected_components_with_threshold(found, min_connections=5)\n",
    "\n",
    "# Keep only islets with at least N cells\n",
    "N = 10\n",
    "large_components = {k: v for k, v in result.items() if len(v) >= N}\n",
    "print(f'Found {len(result)} components, {len(large_components)} components with at least {N} cells..')\n",
    "\n",
    "# Add islet annotations to AnnData\n",
    "mapping = {}\n",
    "for key, cells in large_components.items():\n",
    "    for cell in cells:\n",
    "        assert not cell in mapping, f'Cell {cell} already mapped to {mapping[cell]}.'\n",
    "        mapping[cell] = key\n",
    "ad.obs['Islet'] = ad.obs.index.map(mapping).fillna('Other')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Show all cells in a spatial plot, colored by islet membership"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(ad.obsm['spatial'][:, 0], ad.obsm['spatial'][:, 1], s=0.1, c=ad.obs['Islet'].astype('category').cat.codes, cmap='tab20', alpha=0.5)\n",
    "ax = plt.gca()\n",
    "ax.set_title('Islet Mapping', fontsize=16)\n",
    "ax.set_xlabel('X Coordinate, µm', fontsize=14)\n",
    "ax.set_ylabel('Y Coordinate, µm', fontsize=14)\n",
    "ax.invert_yaxis()\n",
    "ax.set_aspect('equal')\n",
    "x1, y1 = 7500, 8000\n",
    "dx, dy = 1500, 1500\n",
    "plt.plot([x1, x1 + dx, x1 + dx, x1, x1], [y1, y1, y1 + dy, y1 + dy, y1], color='black', linewidth=0.5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Zoom in to the region with islets, plot cells with segmented boundaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x1, y1 = 7500, 8000\n",
    "dx, dy = 1500, 1500\n",
    "plotWithBoundaries(ad, x1, x1 + dx, y1, y1 + dy, s=id, var='Islet', exclusion='',\n",
    "                    dfb=ad.uns['df_cell_boundaries'], figsize=(8, 8), invert=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count number of cell of each endocrine celltype per islet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "se = ad.obs[['Islet', 'Celltype fine']].value_counts()\n",
    "se = se.loc[se.index.get_level_values(0) != 'Other']\n",
    "df = se.unstack().fillna(0).astype(int).T\n",
    "df.loc['Total'] = df.sum()\n",
    "\n",
    "df.drop('Total').loc[:, np.random.choice(df.columns, size=30, replace=False)].style.background_gradient(cmap='Blues', axis=0).format('{:,.0f}').set_properties(**{'text-align': 'center'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_islets_composition_heatmap(df, th=0.1, link_colors=['red', 'teal', 'navy', 'gold', 'brown', 'pink', 'maroon', 'lime'], ct_colors='Blues', figsize=(15, 5)):\n",
    "\n",
    "    \"\"\"Plot a heatmap of islet cell type composition with hierarchical clustering.\n",
    "    This function visualizes the composition of different cell types across islets,\n",
    "    using hierarchical clustering to group similar islets together.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        DataFrame containing cell type composition data for islets.\n",
    "\n",
    "    th : float, optional\n",
    "        Threshold for clustering distance (default is 0.1).\n",
    "\n",
    "    link_colors : list, optional\n",
    "        List of colors for the dendrogram links (default is ['red', 'teal', 'navy', 'gold', 'brown', 'pink', 'maroon', 'lime']).    \n",
    "\n",
    "    ct_colors : list, optional\n",
    "        List of colors for cell types (default is ['Greys', 'Reds', 'Oranges', 'Greens', 'Purples', 'YlOrBr', 'Blues']).\n",
    "    \"\"\"\n",
    "\n",
    "    # Hierarchical clustering for columns (islets)\n",
    "    col_dist = pdist(df.T, metric='correlation')\n",
    "    col_link = linkage(col_dist, method='average')\n",
    "    col_order = dendrogram(col_link, no_plot=True)['leaves']\n",
    "    # Hierarchical clustering for rows (cell types)\n",
    "    row_dist = pdist(df, metric='correlation') \n",
    "    row_link = linkage(row_dist, method='average')\n",
    "    row_order = dendrogram(row_link, no_plot=True)['leaves']\n",
    "\n",
    "    # Reorder data\n",
    "    df_ordered = df.iloc[row_order, col_order]\n",
    "    row_colors = ct_colors[::-1][:len(df_ordered)]\n",
    "    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=figsize, gridspec_kw={'height_ratios': [1, 2]})\n",
    "    hierarchy.set_link_color_palette(link_colors)\n",
    "    dendrogram(col_link, orientation='top', ax=ax1, no_labels=True, color_threshold=th, above_threshold_color='darkgrey')\n",
    "    ax1.set_ylabel('Islets')\n",
    "    ax1.tick_params(labelleft=False)\n",
    "    ax1.axis('off')\n",
    "    ax1.set_title('Islet Cell Type Composition Heatmap', fontsize=16)\n",
    "\n",
    "    ar = df_ordered.loc['Total'].values.reshape(1, -1)\n",
    "    ar = np.sqrt(ar + 1)\n",
    "    ar = ar / int(np.quantile(ar, 0.95))\n",
    "    i = len(df_ordered) - 1\n",
    "    ax2.imshow(ar, cmap='Greens', aspect='auto',\n",
    "            extent=[0, ar.shape[1], i, i+1], vmin=0, vmax=ar.max())\n",
    "\n",
    "    df_non_total = df_ordered.drop('Total')\n",
    "    df_non_total_norm = df_non_total / df_non_total.sum(axis=0)\n",
    "    for i, (idx, row) in enumerate(df_non_total_norm.iterrows()):\n",
    "        ax2.imshow(row.values.reshape(1, -1), cmap=ct_colors, aspect='auto',\n",
    "                extent=[0, len(row), i, i+1], vmin=0, vmax=1)\n",
    "\n",
    "    ax2.set_xticks([])\n",
    "    ax2.set_xticklabels([])\n",
    "    ax2.set_yticks(np.arange(len(df_ordered)) + 0.5)\n",
    "    ax2.set_yticklabels(df_ordered.index[:-1].tolist() + ['Diameter'], fontsize=12)\n",
    "    ax2.set_xlabel('Islets (clustered hierarchically)', fontsize=14)\n",
    "    ax2.set_ylim(0, len(df_ordered))\n",
    "\n",
    "    clusters = np.sort(fcluster(col_link, t=th, criterion='distance'))\n",
    "    locs = np.where((np.roll(clusters, 1) - clusters)<0)[0]\n",
    "    for l in locs:\n",
    "        ax2.axvline(l, color='crimson', linewidth=0.5)\n",
    "\n",
    "    ax2.text(1.01 * df.shape[1], df.shape[0] - 1, 'Darker means\\nlarger', fontsize=14, ha='left')\n",
    "    ax2.text(1.01 * df.shape[1], df.shape[0] / 2, 'Dark is 1\\nLight is 0', fontsize=14, ha='left')\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "    return\n",
    "\n",
    "plot_islets_composition_heatmap(df, th=0.1, link_colors=['brown'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find the largest islet and plot it with boundaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_islet(ad, islet, col='Islet', **kwargs):\n",
    "    ar = ad.obsm['spatial'][ad.obs[col] == islet]\n",
    "    x0, y0 = ar.min(axis=0)\n",
    "    x1, y1 = ar.max(axis=0)\n",
    "    dx, dy = x1 - x0, y1 - y0\n",
    "    b = 0.05\n",
    "    x0 -= dx * b\n",
    "    x1 += dx * b\n",
    "    y0 -= dy * b\n",
    "    y1 += dy * b\n",
    "    dx, dy = x1 - x0, y1 - y0\n",
    "\n",
    "    plotWithBoundaries(ad, x0, x1, y0, y1, s=id, var='Celltype fine', exclusion='NA',\n",
    "                        dfb=ad.uns['df_cell_boundaries'], invert=False, colormap=palette_pancreas_fine,\n",
    "                        **kwargs)\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "islet = df.columns[df.loc['Total'].argmax()]\n",
    "\n",
    "plot_islet(ad, islet, figsize=(6, 6), title=islet)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find the largest beta cell-enriched islets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_proportions = df.drop('Total').div(df.loc['Total'], axis=1)\n",
    "se = df_proportions.loc['Endocrine-Beta'].sort_values(ascending=False)\n",
    "thb = 0.35\n",
    "se = df.loc['Total', se[se>thb].index]\n",
    "se.plot(kind='bar', figsize=(8, 4), color='olive', title=f'Islets with >{thb*100}% Beta cells', legend=False)\n",
    "ax = plt.gca()\n",
    "ax.set_ylabel('Number of Cells In Islet', fontsize=14)\n",
    "ax.set_xlabel('Islet', fontsize=14)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "islet = 'C381'\n",
    "plot_islet(ad, islet, figsize=(2.5, 2.5), title=islet)\n",
    "\n",
    "islet = 'C162'\n",
    "plot_islet(ad, islet, figsize=(2.5, 2.5), title=islet)\n",
    "\n",
    "islet = 'C169'\n",
    "plot_islet(ad, islet, figsize=(2.5, 2.5), title=islet)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## For each islet, find and quantify other cell types that are present in and around the islet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "endocrine_subtypes = ['Endocrine-Alpha', 'Endocrine-Beta', 'Endocrine-Delta', 'Endocrine-Gamma', 'Endocrine-Other-1', 'Endocrine-Other-2']\n",
    "endocrine_cells = ad.obs.index[ad.obs['Celltype fine'].isin(endocrine_subtypes)]\n",
    "found = find_neighbors(ad.obs, radius=16, x_column='x_centroid', y_column='y_centroid', cells_of_interest=endocrine_cells, verbose=True)\n",
    "\n",
    "# Create a copy of the found neighbors to keep the original intact\n",
    "found_all = found.copy()\n",
    "\n",
    "# Filter the found neighbors to only include those that are also endocrine cells\n",
    "for cell in found.keys():\n",
    "    found[cell] = [c for c in found[cell] if c in endocrine_cells]\n",
    "\n",
    "# Find islets as connected components with at least 5 connections\n",
    "result = find_connected_components_with_threshold(found, min_connections=5)\n",
    "\n",
    "# Keep only islets with at least N cells\n",
    "N = 10\n",
    "large_components = {k: v for k, v in result.items() if len(v) >= N}\n",
    "print(f'Found {len(result)} components, {len(large_components)} components with at least {N} cells..')\n",
    "\n",
    "for key in large_components.keys():\n",
    "    cells_in_this_islet = large_components[key]\n",
    "    all_this_cells = []\n",
    "    for c in cells_in_this_islet:\n",
    "        all_this_cells.extend(found_all[c])\n",
    "    large_components[key] = list(set(all_this_cells))\n",
    "\n",
    "# Add islet annotations to AnnData\n",
    "mapping = {}\n",
    "for key, cells in large_components.items():\n",
    "    for cell in cells:\n",
    "        if False:\n",
    "            if cell in mapping:\n",
    "                print(f'Cell {cell} already mapped to {mapping[cell]}.')\n",
    "        mapping[cell] = key\n",
    "ad.obs['Islet-extended'] = ad.obs.index.map(mapping).fillna('Other')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "se = ad.obs[['Islet-extended', 'Celltype fine']].value_counts()\n",
    "se = se.loc[se.index.get_level_values(0) != 'Other']\n",
    "df = se.unstack().fillna(0).astype(int).T\n",
    "df.loc['Total'] = df.sum()\n",
    "\n",
    "plot_islets_composition_heatmap(df, th=0.025, link_colors=['brown'], figsize=(15, 8), ct_colors='magma_r')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
